[server]New parameter for WMS service: tile_buffer

lbartoletti opened the pull request:
## Description
Some applications use tiles (WMS GetMap) and we have a problem with cut symbols.

As for example:
![image](https://user-images.githubusercontent.com/7521540/58852537-41955480-8697-11e9-87f8-59479ebe1552.png)

On openlayer, this can be solved by using the `gutter:` option `in ol.source.TileWMS` But this does not solve another problem for point clusters (another pr will come to fix this after this one)

We propose to implement the equivalent of MapServer's [tile_map_edge_buffer](https://mapserver.org/output/tile_mode.html) option in QGIS Server.

The solution avoids transferring unnecessary pixel volumes to the client side (no need to use gutter).
This requires adding a global parameter to the WMS service as well as a TILED=true parameter read from the GetMap request. Indeed, this option is only taken into account in a tile mode.

Finally with this solution, we have beautiful uncut symbols
![image](https://user-images.githubusercontent.com/7521540/58852839-62aa7500-8698-11e9-8388-30ef88ab594e.png)

Settings for the new parameter:
![image](https://user-images.githubusercontent.com/7521540/58853374-59baa300-869a-11e9-9280-e553d0bdac8a.png)

cc @elemoine @pblottiere @troopa81 

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [X] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [X] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [X] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [X] New unit tests have been added for core changes
- [X] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

lbartoletti committed:
New parameter for WMS service: tile_buffer [needs-docs][FEATURE]

elemoine commented:
This looks great @lbartoletti. I just think we should improve the description text. For example "Tile buffer" in the UI is not explicit enough I think.

m-kuhn commented on src/server/qgsserverprojectutils.h:
```diff
@@ -147,6 +147,14 @@ namespace QgsServerProjectUtils
+   * \returns tile buffer if defined in project, 0 otherwise.
+   * \since QGIS 3.10
+   */
+  SERVER_EXPORT int wmsTileBuffer( const QgsProject &project );
```
const?

lbartoletti replied:
No, it's a non-member function

m-kuhn commented:
I have seen the same happening without tiling on symbols at the edge of the layer bounding box. Can this also solve this issue?

nyalldawson commented on python/server/auto_generated/qgsserverprojectutils.sip.in:
```diff
@@ -152,6 +152,17 @@ Returns the quality for WMS images defined in a QGIS project.
+
+  int wmsTileBuffer( const QgsProject &project );
+%Docstring
+Returns the tile buffer for WMS images defined in a QGIS project.
```
Units? Is this pixels?

elemoine replied:
Fixed.

nyalldawson reviewed (commented):
Wouldn't it make sense to move this setting up QgsMapSettings and handle it in the map render jobs? I think there's a strong use case for this outside of server too. (Eg the new XYZ tile creation algorithm)

troopa81 commented on src/server/services/wms/qgswmsrendercontext.cpp:
```diff
@@ -638,22 +648,40 @@ bool QgsWmsRenderContext::isValidWidthHeight() const
+    throw QgsBadRequestException( QgsServiceException::QGIS_InvalidParameterValue,
+                                  mParameters[QgsWmsParameter::BBOX] );
+  }
+  const double extentWidth = extent.width();
```
Is const really necessary here? and I would also recommend to remove the intermediate extentheight, extentWidth, width, height variables

andreasneumann commented:
> I have seen the same happening without tiling on symbols at the edge of the layer bounding box. Can this also solve this issue?

Yes - I can second this. In Kanton Zug we also had issues at the edges of a layer overall extent. Point-Symbols are only partially drawn (the part inside the bounding box) and most likely, if one had a thicker line-width at the layer boundary, this would be cut off as well ;-(

lbartoletti committed:
fix qgis_spelling

lbartoletti committed:
troopa's style

lbartoletti committed:
silly and unmodified copy/paste

wonder-sk commented:
> Wouldn't it make sense to move this setting up QgsMapSettings and handle it in the map render jobs? I think there's a strong use case for this outside of server too. (Eg the new XYZ tile creation algorithm)

Yes please! And even when rendering map in QGIS desktop one can spot this behavior.

Implementation-wise, in addition if this is done in QgsMapSettings, we could just use a bit bigger source extent for getFeatures() while keeping the destination image of the same size. In the current implementation on this PR, things seem a bit suboptimal, because the rendered map image is larger, and then a region is cut out for the final image. And with the current approach, I think there would be also issues with labels being cut off at the edges of tiles because of larger map image size...?

elemoine commented:
Thank you all for your comments on this PR. I am also involved in this work on "cut symbols at tile boundaries", and I'm going to respond to the comments and questions that have been raised.

> > I have seen the same happening without tiling on symbols at the edge of the layer bounding box. Can this also solve this issue?
> 
> Yes - I can second this. In Kanton Zug we also had issues at the edges of a layer overall extent. Point-Symbols are only partially drawn (the part inside the bounding box) and most likely, if one had a thicker line-width at the layer boundary, this would be cut off as well ;-(

I think this is a separate issue. The "issue" you are referring to is related to this [thread](https://lists.osgeo.org/pipermail/qgis-developer/2019-May/thread.html#57225) initiated by @andreasneumann  on the QGIS-Developer mailing list. And please see the [answer](https://lists.osgeo.org/pipermail/qgis-developer/2019-May/057237.html) I provided in the thread, which @elpaso [agreed](https://lists.osgeo.org/pipermail/qgis-developer/2019-May/057238.html) with :)

The issue addressed with this PR is related to the use of **tiles**, and symbols (and thick lines) cut at the boundaries of **tiles**.

>> Wouldn't it make sense to move this setting up QgsMapSettings and handle it in the map render jobs? I think there's a strong use case for this outside of server too. (Eg the new XYZ tile creation algorithm)
>
> Yes please! And even when rendering map in QGIS desktop one can spot this behavior.

Yep, I agree with @nyalldawson and @wonder-sk that we could possibly move this outside the server, to make it available to the new XYZ tile creation algorithm. We'll have a look at that.

> Implementation-wise, in addition if this is done in QgsMapSettings, we could just use a bit bigger source extent for getFeatures() while keeping the destination image of the same size. In the current implementation on this PR, things seem a bit suboptimal, because the rendered map image is larger, and then a region is cut out for the final image.

I just gave this a try, that is using a bigger extent for getFeatures but keeping the destimation image of the same size. And it doesn't work – I observe cut symbols at the tile boundaries again. I suppose that points that are geographically outside the image are not rendered at all, although I haven't found where this exclusion actually occurs.

elemoine commented:
> > Implementation-wise, in addition if this is done in QgsMapSettings, we could just use a bit bigger source extent for getFeatures() while keeping the destination image of the same size. In the current implementation on this PR, things seem a bit suboptimal, because the rendered map image is larger, and then a region is cut out for the final image.
> 
> I just gave this a try, that is using a bigger extent for getFeatures but keeping the destimation image of the same size. And it doesn't work – I observe cut symbols at the tile boundaries again. I suppose that points that are geographically outside the image are not rendered at all, although I haven't found where this exclusion actually occurs.

@wonder-sk I did more tests (and more fixes to my code :) and you were right, it works, we just need to use a bigger source extent, while keeping the destination image of the same size.  I am working on a new set of changes that take all that into account.  Thanks again.

Commit:
Keep the destination image of the same size

Use a bigger extent for getFeatures but keep the destination image of the
same size.

Commit:
Update API docs for wmsTileBuffer

m-kuhn commented on src/core/qgsmapsettings.h:
```diff
@@ -102,6 +102,16 @@ class CORE_EXPORT QgsMapSettings
+    double extentBuffer() const;
+
+    /**
+     * \since QGIS 3.10
```
This will need some more docstrings to explain what this does in detail

elemoine replied:
Sure. This is still WIP ;)

elemoine replied:
Done.

nyalldawson replied:
Can you also describe what units this is in?

elemoine replied:
Good call. New commit added for this.

elemoine replied:
And for the record I also added a tooltip to describe the TileBuffer setting in the UI.

m-kuhn commented on src/server/services/wms/qgswmsrendercontext.cpp:
```diff
@@ -648,36 +649,42 @@ bool QgsWmsRenderContext::isValidWidthHeight() const
 }
 
-QgsRectangle QgsWmsRenderContext::mapExtent() const
+double QgsWmsRenderContext::mapTileBuffer(int mapWidth) const
```
Consider setting up the pre-commit hook to make your life easier ;)

elemoine replied:
Yep, done.

m-kuhn reviewed (commented):
Looking good :+1: 
Did you plan to write some unit tests for the `extentBuffer()` too?

Commit:
Improve the API docs

Commit:
Reformat mapTileBuffer signature

Commit:
Add a unit test for QgsMapSettings::extentBuffer

Commit:
Set extent buffer in configureMapSettings

Commit:
Add missing const's

elemoine commented:
> Did you plan to write some unit tests for the `extentBuffer()` too?

Yes, I added a test showing how to use `visibleExtent()` and `extentBuffer()` to create a buffered extent.

Commit:
Fix code styling

elemoine commented:
Travis is green! I guess this is time for a new review round :) Any new comment or question is very welcome. For your information we've been testing this with an OpenLayers map, and it's working great.

Commit:
Add tooltip for the TileBuffer setting

Commit:
Improve the extentBuffer API docs

nyalldawson commented:
Nice changes!

One question -- how does this relate to the existing QgsFeatureRenderer::modifyRequestExtent option? I wonder if we could avoid the need for manually setting parameter in certain circumstances by using modifyRequestExtent to automatically expand the request extent to account for large symbols.

elemoine commented:
> One question -- how does this relate to the existing QgsFeatureRenderer::modifyRequestExtent option?

Good call again! I'll have a look at that.

elemoine commented:
> One question -- how does this relate to the existing QgsFeatureRenderer::modifyRequestExtent option? I wonder if we could avoid the need for manually setting parameter in certain circumstances by using modifyRequestExtent to automatically expand the request extent to account for large symbols.

@nyalldawson I don't think this can work. For example think about the case where the symbol size is the result of the evaluation of an expression involving a feature attribute. In that case it is not possible for modifyRequestExtent to determine how to extend the request extent. So, IMO, relying on a user-defined setting for the "tile buffer" makes for a simple solution without adding complexity to the QGIS code base. And I think it's also easy enough for users – if they still see cut symbols they can just make the tile buffer a bit larger to really take the bigger symbols into account.

nyalldawson commented:
@elemoine ok, fair point!

pblottiere commented:
Hi @lbartoletti and @elemoine,

Thanks a lot for your work, I'm very happy to see this feature landing in QGIS Server :+1:!

Code looks good now, and I'll merge this PR tomorrow if I we don't have comments in the meantime.

And thanks @m-kuhn and @nyalldawson for reviewing this PR, Eric is whispering to me that it was a true pleasure :).

elpaso commented:
@lbartoletti nice! Thank you for addressing this issue.

lbartoletti commented:
> @lbartoletti nice! Thank you for addressing this issue.

@elpaso Thank you. But the credit goes to the whole Oslandia team with the support of the QGIS community for the review :wink:

elpaso commented:
> > @lbartoletti nice! Thank you for addressing this issue.
> 
> @elpaso Thank you. But the credit goes to the whole Oslandia team with the support of the QGIS community for the review wink

Kudos to the whole Oslandia team then! And "live long and prosper" to all QGIS Team :wink:

andreasneumann commented:
Will this be backported to 3.4 LT or is this too much of a change?

I guess most QGIS server users will remain at version 3.4 for quite some time.

lbartoletti commented:
> Will this be backported to 3.4 LT or is this too much of a change?

@andreasneumann I think it could be backported easily.

nyalldawson commented:
Yeah, I think this can be backported safely too

pblottiere commented:
> Will this be backported to 3.4 LT or is this too much of a change?

Considering the fact that this PR introduces a completely new parameter, the basic behavior of QGIS Server still remain the same.

So +1 for backporting this PR too.

elemoine commented:
And also, I'll write some documentation for this.

rldhont commented:
Is this PR backported ?

lbartoletti commented:
On reflection, as it is almost a feature and given the refacto that there was (especially QgsWmsRenderContext) between 3.4 and 3.8, I did not continue the manual backport.

mbernasocchi commented:
@lbartoletti so no backport for 3.4. Do you still want it in  3.8? backport bot did not even try backporting to 3.8. it seems that once a backport fails, it does not try the other. I had the same issue in a PR before and removing and readding the `backport release-3_8` tag made it go through.

karl-magnus commented:
Hi! Trying this out now. Works nice for symbols. But how about labels? Seems not to effect them. Are there only external solutions to fix such issues? Or is it possible to include them too?
![bild](https://user-images.githubusercontent.com/4749245/83875120-d7ded900-a736-11ea-9bba-55fc2a4a26eb.png)

troopa81 commented:
@karl-magnus label position are computed automatically so they don't overlap, but if you get several tiles, you are not garanteed to have the same place on every tile and you end up in this situation. 

You can solve this is by fixing your label position
![poslabel](https://user-images.githubusercontent.com/14358135/83877434-d7e0d800-a73a-11ea-9aa4-00d44ce20fda.png)

hope it helps!

karl-magnus commented:
@troopa81 Thanks for answering. I already use fixed positions with data defined position, alignment and rotation. If it wasn't fixed I would probably have got two labels, one on each side of the tile border. I would expect this setting to force the creation of a, in my case, 60 px bigger picture than the wms request in each direction. Then it would be cropped to requested dimensions after rendering resulting in that the label would be rendered at the same place in both adjacent tiles. It looks like the label in my example isn't rendered in the upper image just the other.

troopa81 commented:
> I would expect this setting to force the creation of a, in my case, 60 px bigger picture than the wms request in each direction. 

@karl-magnus I would also expect it. I'm not sure if this is done on purpose or not but the label job doesn't [use the extentBuffer](https://github.com/qgis/QGIS/blob/master/src/core/qgsmaprendererjob.cpp#L584) but only the visible extent.

Could you create an issue with test project and data sample and mention me. I'll try to fix it when I get the time.
